tests/cluster/tests/24-links.tcl

```diff
@@ -0,0 +1,99 @@
+            set to 0
+            set from 0
+            foreach l $links {
+                if {[dict get $l node] eq $peer} {
+                    if {[dict get $l dir] eq "to"} {
+                        incr to
+                    } elseif {[dict get $l dir] eq "from"} {
+                        incr from
+                    }
+                }
+            }
+            assert {$to eq 1}
+            assert {$from eq 1}
+        }
+    }
+}
+
+set primary1_id 0
+set primary2_id 1
+
+set primary1 [Rn $primary1_id]
+set primary2 [Rn $primary2_id]
+
+test "Disconnect link when send buffer limit reached" {
+    # On primary1, set timeout to 1 hour so links won't get disconnected due to timeouts
+    set oldtimeout [lindex [$primary1 CONFIG get cluster-node-timeout] 1]
+    catch {$primary1 CONFIG set cluster-node-timeout [expr 60*60*1000]}
+
+    # Get primary1's links with primary2
+    set primary2_name [dict get [get_myself $primary2_id] id]
+    set orig_link_p1_to_p2 [get_link_to_peer $primary1_id $primary2_name]
+    set orig_link_p1_from_p2 [get_link_from_peer $primary1_id $primary2_name]
+
+    # On primary1, set cluster link send buffer limit to 32MB
+    set oldlimit [lindex [$primary1 CONFIG get cluster-link-sendbuf-limit] 1]
+    catch {$primary1 CONFIG set cluster-link-sendbuf-limit [expr 32*1024*1024]}
+
+    # To manufacture an ever-growing send buffer from primary1 to primary2,
+    # make primary2 unresponsive.
+    set primary2_pid [get_instance_attrib redis $primary2_id pid]
+    exec kill -SIGSTOP $primary2_pid
+
+    # On primary1, send a 10MB Pubsub message. It will stay in send buffer of
+    # the link from primary1 to primary2
+    catch {$primary1 publish channel [prepare_value [expr 10*1024*1024]]}
+
+    # Check the same link has not been disconnected, but its send buffer has grown
+    set same_link_p1_to_p2 [get_link_to_peer $primary1_id $primary2_name]
+    assert {[dict get $same_link_p1_to_p2 ctime] eq [dict get $orig_link_p1_to_p2 ctime]}
+    assert {[dict get $same_link_p1_to_p2 sbufalloc] > [dict get $orig_link_p1_to_p2 sbufalloc]}
+
+    # On primary1, send another 30MB Pubsub message.
+    catch {$primary1 publish channel [prepare_value [expr 30*1024*1024]]}
+    
+    # Link has exceeded buffer limit and been dropped and recreated
+    set new_link_p1_to_p2 [get_link_to_peer $primary1_id $primary2_name]
+    assert {[dict get $new_link_p1_to_p2 ctime] > [dict get $orig_link_p1_to_p2 ctime]}
+
+    # Link from primary2 should not be affected
+    set same_link_p1_from_p2 [get_link_from_peer $primary1_id $primary2_name]
+    assert {[dict get $same_link_p1_from_p2 ctime] eq [dict get $orig_link_p1_from_p2 ctime]}
+
+    # Revive primary2
+    exec kill -SIGCONT $primary2_pid
+
+    # Reset configs on primary1 so config changes don't leak out to other tests
+    catch {$primary1 CONFIG set cluster-node-timeout $oldtimeout}
+    catch {$primary1 CONFIG set cluster-link-sendbuf-limit $oldlimit}
```

Comment:
Why are we catching these exceptions? Also if the test fails midway through these changes will get lost.

Reply:
Good question. I was copying what existing tests did. Looking at it now, it makes more sense to let the exceptions be raised and fail the test, instead of catching and dropping them.

Re: tests failing midway through. If one test case fail, does the test framework continue to run the remaining tests? If yes, this sounds like the generic issue, i.e. the lack of isolation between cluster tests. How do existing tests handle prevent "leaks"?

Reply:
Yeah, the testing framework is a little bit silly in that it doesn't do a full reset of the cluster nodes.